substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/graal/code/SubstrateBackend.java

```diff
@@ -152,14 +166,220 @@ public CompilationResult newCompilationResult(CompilationIdentifier compilationI
         return new CompilationResult(compilationIdentifier, name) {
             @Override
             public void close(OptionValues options) {
-                /*
-                 * Do nothing, we do not want our CompilationResult to be closed because we
-                 * aggregate all data items and machine code in the native image heap.
-                 */
+                FrameOptimizer.compressInfoPoints(this);
             }
         };
     }
 
+    private static class FrameOptimizer {
+        /**
+         * Wrapper class that makes up for deficiencies in the implementation of
+         * BytecodeFrame.equals.
+         */
+        private static class BytecodeFrameComparator {
```

Comment:
This is a very unusual implementation of a `Comparator` in Java. Usually you implement a `java.util.Comparator` class specialized to the type you want to compare. E.g. see `com.oracle.svm.hosted.meta.HostedUniverse.TypeComparator`. Then you have one comparator instance (and not one for each `BytecodeFrame`) that you use with the two `BytecodeFrames` that you want to compare.

Reply:
Since the only state you have in `BytecodeFrameComparator` is the `BytecodeFrame` you want to compare against I suggest you choose an implementation that does not require creating a `BytecodeFrameComparator` for each `BytecodeFrame` you want to compare against.

Also the term Comparator here is misleading because you only care about equality and hash-code so this isn't a Comparator in the sense of `java.util.Comparator`. I suggest replacing this class with static methods that take two `BytecodeFrame` instances. And in the `frameMap` you have two `BytecodeFrames` instead.

Reply:
Yes, this is not actually a Comparator in the sense of `java.util.Comparator`. It was just a wrapper type so I could redefine the equals and hash methods.

I see that `EconomicMap.create` accepts an argument of type `Equivalence` to define the `equals` and `hash` operations used to manage map inserts and lookups (the map is actually being used as a hashset here). As you say, I believe all I need to do here is implement `Equivalence` appropriately with, say, class `BytecodeFrameEquivalence` and then create an EconomicMap<BytecodeFrame,BytecodeFrame> parameterised passing an instance of `BytecodeFrameEquivalence` to the create call. I will only ever need one instance across all uses.